src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1499 @@ void Init()
+    {
+        switch (this)
+        {
+            case { IsOk: false }:
+                return Value;
+            default:
+                return Value; // 1
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (18,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(18, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_WithNestedDeclaration()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { IsOk: true, Value: var value }:
+                return value;
+            default:
+                return Value; // 1
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (18,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(18, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_WithNestedDeclaration_ReverseOrder_Struct()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { Value: var value, IsOk: true }:
+                return value;
+            case { Value: var value }:
+                return value; // 1
+            default:
+                return Value; // 2
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (18,24): warning CS8603: Possible null reference return.
+                //                 return value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "value").WithLocation(18, 24),
+                // (20,17): warning CS0162: Unreachable code detected
+                //                 return Value; // 2
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "return").WithLocation(20, 17)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_SideEffectOnStruct()
```

Comment:
>MemberNotNullWhenTrue_RecursivePattern_SideEffectOnStruct [](start = 20, length = 57)

Is this test needed? And where is the side-effect?

Reply:
This test illustrates that we're not testing a copy of the struct, but the original. Although this doesn't reflect the lowered code, we can do that because analysis of patterns doesn't care about properly understanding side-effects.

---
In reply to: [573108996](https://github.com/dotnet/roslyn/pull/51001#discussion_r573108996) [](ancestors = 573108996)